### ISSUE TITLE
Upgrading from 2.3.3 to 3.0.1 breaks stylesheet @import

### ISSUE THREAD

**Issue by johnlane** [rails/sprockets-rails#324]
Having just updated an app from Rails 4.2.4 to 4.2.5.1, one of my SASS stylesheets has stoped working, reporting an "undefined variable". 

The variable in question is defined by the import of another file that happens before it, by my `application.scss` which is along these lines:

    @import "bootstrap";
    @import "bootstrap-social";

where the latter file references a varable defined by the former. The latter file is an underscore-prefixed file `_bootstrap-social.scss`.

Prior to the update, everything worked. I resolved the issue by specifically reqesting `gem 'sprockets-rails', '2.3.3'` in my `Gemfile`. Without that specificity I get version 3.0.1 of `sprockets-rails`and the aforementioned undefined variable issue.

I can make it work with 3.0.1 by adding `@import 'bootstrap` to the head of `_boootstrap-social.scss` but this shouldn't be necessary (and isn't wth 2.3.3).

**Comment by rafaelfranca** [comment]
Can you please provide a sample application that reproduces the error?

**Comment by johnlane** [comment]
Please see https://github.com/johnlane/test-sprockets-rails. There are two branches:

* `test-rails-4.2.4` originated with Rails 4.2.4 and was upgraded to Rals 4.2.5.1,
* `test-rails-4.2.5.1` originated with Rails 4.2.5.1 and remained so.

There are actually two problems. The first affects both branches and occurs when `require tree .` is used in the stylesheet. I don't normally use a wildcard require like that, so only noticed this when investigating the reported problem. 

I discovered that the second problem is only an issue on the branch that started life as Rails 4.2.4 and was later upgraded to Rails 4.2.5.1 and that is that the variable visibility of `@import` is affected by the version of `sprockets-rails`. 

* The test app, with Rails 4.2.4, later upgraded to Rails 4.2.5.1 does not work with sprockets-rails version 3.0.1 but it does wotk with 2.3.3. Providing of course that `require tree .` is not used.

* The test app, with Rails 4.2.5.1 works with sprockets-rails version 3.0.1 or 2.3.3 as long as `require tree .` is not used.

My real app, where I encountered the problem first, started out on Rails 4.2.4 and was recently upgraded to Rails 4.2.5.1.

Hope that makes sense. See the commit histories on the tests for more idea of what I did.

**Comment by rafaelfranca** [comment]
You should not mix `require` and `@import`. See https://github.com/rails/sass-rails/pull/280

**Comment by johnlane** [comment]
Hi you closed the issue but I think missed the point. It isn't about require mixed with import. I don't do that in my app. It's about an app on or upgraded from rails 4.2.4 not working with sprockets-rails 3.0.1.

**Comment by mattbrictson** [comment]
> It isn't about require mixed with import. I don't do that in my app.

But your app *does* do that, right? Your application.scss has both `*= require_self` and `@import` in the same file. It sounds like @rafaelfranca is saying that mixing these is no longer supported.

**Comment by johnlane** [comment]
I will look at that tomorrow. It might have require self in there.

**Comment by rafaelfranca** [comment]
If you app don't do that so please change the example application to reproduce the error in the same way as your application.

**Comment by johnlane** [comment]
Ok, so I had a `require_self` in there (it didn't serve any purpose, it was just there since Rails generated the original file and I never removed it). Once I removed it, the problem went away.

However, the presence of this wasn't an issue when I was using `sprockets-rails` 2.3.3. it only became an issue when I upgraded my app from Rails 4.2.4. I thought it was interesting that an app that originated on Rails 4.2.5.1 did not suffer this problem and worked as expected with `sprockets-rails` 3.0.1 despite having `require_self` in there.

I have, however, resolved the problem for myself by not using Sprockets `require` in SASS files so, from my point of view, this can be closed.
